File: critter/core/src/test/java/dev/morphia/critter/parser/TestAccessorsMutators.java
package dev.morphia.critter.parser;

import java.util.List;

import dev.morphia.critter.Critter;
import dev.morphia.critter.CritterClassLoader;
import dev.morphia.critter.CritterKt;
import dev.morphia.critter.sources.Example;

import org.bson.codecs.pojo.PropertyAccessor;
import org.testng.Assert;
import org.testng.annotations.DataProvider;

public class TestAccessorsMutators extends BaseCritterTest {
    private final CritterClassLoader critterClassLoader = new CritterClassLoader();

    // @Test(dataProvider = "classes")
    public void testPropertyAccessors(Class<?> type) throws Exception {
        List<List<Object>> testFields = List.of(
                List.of("name", String.class, "set externally"),
                List.of("age", int.class, 100),
                List.of("salary", Long.class, 100_000L));

        Object entity = critterClassLoader.loadClass(type.getName()).getConstructor().newInstance();

        for (List<Object> field : testFields) {
            testAccessor(type, critterClassLoader, entity, (String) field.get(0), field.get(2));
        }
    }

    @SuppressWarnings("unchecked")
    private void testAccessor(
            Class<?> type,
            CritterClassLoader loader,
            Object entity,
            String fieldName,
            Object testValue) throws Exception {
        Class<PropertyAccessor<Object>> accessorClass = (Class<PropertyAccessor<Object>>) loader.loadClass(
                Critter.Companion.critterPackage(type)
                        + type.getSimpleName()
                        + CritterKt.titleCase(fieldName)
                        + "Accessor");
        PropertyAccessor<Object> accessor = accessorClass.getConstructor().newInstance();

        accessor.set(entity, testValue);
        Assert.assertEquals(accessor.get(entity), testValue);
        Assert.assertTrue(
                entity.toString().contains(testValue.toString()),
                "Could not find '" + testValue + "` in :" + entity);
    }

    @DataProvider(name = "classes")
    public Object[][] names() {
        return new Object[][] { { Example.class } };
    }
}

File: critter/core/src/test/kotlin/dev/morphia/critter/parser/TestAccessorsMutators.kt
package dev.morphia.critter.parser

import dev.morphia.critter.Critter.Companion.critterClassLoader
import dev.morphia.critter.Critter.Companion.critterPackage
import dev.morphia.critter.parser.java.CritterClassLoader
import dev.morphia.critter.sources.Example
import dev.morphia.critter.titleCase
import org.bson.codecs.pojo.PropertyAccessor
import org.testng.Assert.assertEquals
import org.testng.Assert.assertTrue
import org.testng.annotations.DataProvider

class TestAccessorsMutators : BaseCritterTest() {
    //    @Test(dataProvider = "classes")
    fun testPropertyAccessors(type: Class<*>) {
        val testFields =
            listOf(
                listOf("name", String::class.java, "set externally"),
                listOf("age", Int::class.java, 100),
                listOf("salary", java.lang.Long::class.java, 100_000L)
            )

        val entity = critterClassLoader.loadClass(type.name).getConstructor().newInstance()

        testFields.forEach { field ->
            testAccessor(type, critterClassLoader, entity, field[0] as String, field[2])
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun testAccessor(
        type: Class<*>,
        critterClassLoader: CritterClassLoader,
        entity: Any,
        fieldName: String,
        testValue: Any,
    ) {
        val accessor =
            (critterClassLoader.loadClass(
                    "${critterPackage(type)}${type.simpleName}${fieldName.titleCase()}Accessor"
                ) as Class<PropertyAccessor<Any>>)
                .getConstructor()
                .newInstance()

        accessor.set(entity, testValue)

        assertEquals(accessor.get(entity), testValue)
        assertTrue(
            entity.toString().contains(testValue.toString()),
            "Could not find '$testValue` in :${entity}"
        )
    }

    @DataProvider(name = "classes")
    fun names(): Array<Class<out Any>> {
        return arrayOf(Example::class.java /*, KotlinDummyEntity::class.java*/)
    }
}
